File: solution/java/day22/src/main/java/diamond/Diamond.java
package diamond;

import io.vavr.collection.Seq;
import io.vavr.control.Option;

import static io.vavr.API.Some;
import static io.vavr.collection.Stream.range;
import static io.vavr.control.Option.none;

public class Diamond {
    public static final char START = 'A';

    private Diamond() {
    }

    public static Option<String> print(char endCharacter) {
        return isValidCharacter(endCharacter)
                ? Some(fullDiamondSafely(endCharacter))
                : none();
    }

    private static String fullDiamondSafely(char endCharacter) {
        return concatLines(
                generateDiamond(endCharacter)
        );
    }

    private static Seq<String> generateDiamond(char endCharacter) {
        var halfDiamond = generateHalfDiamond(endCharacter);
        return halfDiamond.appendAll(
                halfDiamond.reverse().drop(1)
        );
    }

    private static Seq<String> generateHalfDiamond(char endCharacter) {
        return range(START, START + (endCharacter - START) + 1)
                .map(i -> (char) i.intValue())
                .map(c -> toLine(c, endCharacter));
    }

    private static String toLine(char character, char endCharacter) {
        var out = outer(character, endCharacter);
        return out + character + inner(character) + out;
    }

    private static String outer(char character, char endCharacter) {
        return generateEmptyCharacters(endCharacter - character);
    }

    private static String inner(char character) {
        return character != START
                ? generateEmptyCharacters(numberOfEmptyCharactersFor(character)) + character
                : "";
    }

    private static int numberOfEmptyCharactersFor(char character) {
        return (character - START) * 2 - 1;
    }

    private static String generateEmptyCharacters(int count) {
        return range(0, count)
                .foldLeft("", (acc, c) -> acc + " ");
    }

    private static boolean isValidCharacter(char endCharacter) {
        return endCharacter >= START && endCharacter <= 'Z';
    }

    private static String concatLines(Seq<String> lines) {
        return lines.mkString(System.lineSeparator());
    }
}

File: solution/kotlin/day22/src/main/kotlin/diamond/Diamond.kt
package diamond

import arrow.core.None
import arrow.core.Option
import arrow.core.Some

object Diamond {
    private const val START = 'A'

    fun Char.toDiamond(): Option<String> = print(this)

    private fun print(endCharacter: Char): Option<String> =
        if (isValidCharacter(endCharacter)) Some(fullDiamondSafely(endCharacter)) else None

    private fun fullDiamondSafely(endCharacter: Char): String =
        concatLines(
            generateDiamond(endCharacter)
        )

    private fun generateDiamond(endCharacter: Char): List<String> =
        generateHalfDiamond(endCharacter)
            .let { halfDiamond ->
                return halfDiamond + halfDiamond.reversed().drop(1)
            }

    private fun generateHalfDiamond(endCharacter: Char): List<String> =
        (START.code..START.code + (endCharacter.code - START.code))
            .map { it.toChar() }
            .map { c -> toLine(c, endCharacter) }

    private fun toLine(character: Char, endCharacter: Char): String =
        outer(character, endCharacter)
            .let { out ->
                return out + character + inner(character) + out
            }

    private fun outer(character: Char, endCharacter: Char): String =
        generateEmptyCharacters(endCharacter.code - character.code)

    private fun inner(character: Char): String =
        if (character != START) generateEmptyCharacters(numberOfEmptyCharactersFor(character)) + character else ""

    private fun numberOfEmptyCharactersFor(character: Char): Int = (character.code - START.code) * 2 - 1

    private fun generateEmptyCharacters(count: Int): String = (0..<count)
        .fold("") { acc, _ -> "$acc " }

    private fun isValidCharacter(endCharacter: Char): Boolean = endCharacter in START..'Z'

    private fun concatLines(lines: List<String>): String = lines.joinToString(System.lineSeparator())
}
